### ISSUE TITLE
Deterministic ECDSA Signatures (RFC 6979)

### ISSUE THREAD

**Issue by paragonie-scott** [phpecc/phpecc#229]
Ref: https://tools.ietf.org/html/rfc6979

To prevent k-value reuse (or even k-value *bias* which could leak the signing key), I propose an interface that looks like this:

```php
interface DeterministicECDSA {
    public function setCurve(NamedCurve $curve); // optional, if this interface isn't attached to a curve class
    public function easySign(string $message, PrivateKey $key): string;
    public function easyVerify(string $message, PublicKey $key, string $signature): bool;
}
```

**Comment by afk11** [comment]
Good idea! My preference has always been to eventually do this.. I'd agree if you said the Signer::sign method isn't set up for success with this interface:

https://github.com/phpecc/phpecc/blob/master/src/Crypto/Signature/Signer.php#L36, but instead we need a KGeneratorInterface, which can differ between random and deterministic implementations? Pretty sure high level sign method would still have enough information to support RFC6979 by default.

**Comment by paragonie-scott** [comment]
As a minor update to this issue: I released [Easy-ECC](https://github.com/paragonie/easy-ecc) a while ago.

Easy-ECC is a usability wrapper which defaults to libsodium (or sodium_compat).
